python/plugins/processing/core/parameters.py

```diff
@@ -79,13 +80,25 @@ def setValue(self, obj):
         of parameter.
         """
         if obj is None:
-            if not self.optional:
+            if not self.optional and self.default is None:
                 return False
-            self.value = None
-            return True
+            elif self.default is None:
+                self.value = None
+                return True
+            else:
```

Comment:
What does this do?

    param = Parameter(optional=True, default='foo')
    print(param.value)

And what this?

    param = Parameter(optional=False, default='foo')
    print(param.value)